src/nfa/nfa.cc

```diff
@@ -623,49 +623,34 @@ bool Mata::Nfa::are_state_disjoint(const Nfa& lhs, const Nfa& rhs)
     return true;
 } // are_disjoint }}}
 
-void Mata::Nfa::make_complete(
+bool Mata::Nfa::make_complete(
         Nfa&             aut,
         const Alphabet&  alphabet,
         State            sink_state)
 {
-    std::list<State> worklist(aut.initial.begin(),
-                              aut.initial.end());
-    std::unordered_set<State> processed(aut.initial.begin(),
-                                        aut.initial.end());
-
-    if (aut.size() <= sink_state) {
-        aut.add_state(sink_state);
-    }
-
-    worklist.push_back(sink_state);
-    processed.insert(sink_state);
-    while (!worklist.empty())
-    {
-        State state = *worklist.begin();
-        worklist.pop_front();
+    bool was_something_added = false;
 
+    auto num_of_states = aut.size();
+    for (State state = 0; state < num_of_states; ++state) {
         std::set<Symbol> used_symbols;
-        if (!aut.delta.empty())
-        {
-            for (const auto &symb_stateset: aut[state]) {
-                // TODO: Possibly fix insert.
-                used_symbols.insert(symb_stateset.symbol);
-
-                const StateSet &stateset = symb_stateset.targets;
-                for (const auto &tgt_state: stateset) {
-                    bool inserted;
-                    tie(std::ignore, inserted) = processed.insert(tgt_state);
-                    if (inserted) { worklist.push_back(tgt_state); }
-                }
-            }
+        for (auto const &tran : aut.delta[state]) {
+            used_symbols.insert(tran.symbol);
         }
-
         auto unused_symbols = alphabet.get_complement(used_symbols);
         for (Symbol symb : unused_symbols)
         {
             aut.delta.add(state, symb, sink_state);
+            was_something_added = true;
+        }
+    }
+
+    if (was_something_added && num_of_states <= sink_state) {
+        for (Symbol symbol : alphabet.get_alphabet_symbols()) {
+            aut.delta.add(sink_state, symbol, sink_state);
```

Comment:
Not sure about the context, but adding transitions by delta.add might be slow, one needs to search and maybe insert the move with the symbol, and insert the target state.

Reply:
I think `get_alphabet_symbols()` should give you the symbols ordered and `add` should first check if the newly added transition should not go to end, so it should be efficient. @Adda0 is it how it works?

Reply:
> I think get_alphabet_symbols() should give you the symbols ordered and add should first check if the newly added transition should not go to end, so it should be efficient. @Adda0, is it how it works?

That is right. This is equivalent to pushing to the back directly.

Reply:
You mean by the problem that it is a bit inefficient? That maybe we should take as input the ordered vector of symbols already, so if you want to use complement multiple times, you can compute the set of symbols directly?
Or maybe it is better for `OnTheFlyAlphabet` to keep some ordered vector of symbols and return that in `get_alphabet_symbols()`?

Reply:
> You mean by the problem that it is a bit inefficient?

Yes, just a little, but still.

It is however called only once in complement, so I would not spend our energy on this as of now.

> if you want to use complement multiple times, you can compute the set of symbols directly?

We could add an additional optional parameter for a set of symbols, but I do not think there will be a use case for that. What do you think?

Reply:
I can imagine use case, you can have alphabet that is used by many automata, and you are doing complement for each of them. I even think we are doing that in noodler, right?

Reply:
We will create an issue from this to resolve later on.